src/neurostatslib/utils.py

```diff
@@ -4,44 +4,263 @@
 from __future__ import annotations
 
 from functools import partial
-from typing import Optional, Union
+from typing import Iterable, List, Optional, Union
 
 import jax
 import jax.numpy as jnp
 import matplotlib.pyplot as plt
 import numpy as np
-from numpy.typing import ArrayLike
+from numpy.typing import ArrayLike, NDArray
 
-# Broadcasted 1d convolution operations.
-# [[n x t],[w]] -> [n x (t - w + 1)]
+# Same trial duration
+# [[r x t x n], [w]] -> [r x (t - w + 1) x n]
+# Broadcasted 1d convolution operations
 _CORR1 = jax.vmap(partial(jnp.convolve, mode="valid"), (0, None), 0)
+_CORR2 = jax.vmap(_CORR1, (2, None), 2)
+_CORR_SAME_TRIAL_DUR = jax.vmap(_CORR2, (None, 1), 3)
+
+# Variable trial dur
 # [[n x t],[p x w]] -> [n x p x (t - w + 1)]
-_CORR2 = jax.vmap(_CORR1, (None, 0), 0)
+_CORR3 = jax.vmap(partial(jnp.convolve, mode="valid"), (1, None), 1)
+_CORR_VARIABLE_TRIAL_DUR = jax.vmap(_CORR3, (None, 1), 2)
+
 
+def check_dimensionality(
+    iterable: Union[NDArray, Iterable[NDArray], jnp.ndarray, Iterable[jnp.ndarray]],
+    expected_dim: int,
+) -> bool:
+    """
+    Check the dimensionality of the arrays in iterable.
 
-def convolve_1d_basis(basis_matrix: ArrayLike, time_series: ArrayLike) -> ArrayLike:
-    """Parameters
+    Check that all arrays in iterable has the expected dimensionality.
+
+    Parameters
     ----------
-    basis_matrix : (n_basis_funcs, window_size)
-        Matrix holding 1d basis functions.
-    time_series : (n_neurons, n_timebins)
-        Matrix holding multivariate time series.
+    iterable :
+    Array-like object containing numpy or jax.numpy NDArrays.
+    expected_dim :
+    Number of expected dimension for the NDArrays.
```

Comment:
```suggestion
    iterable :
        Array-like object containing numpy or jax.numpy NDArrays.
    expected_dim :
        Number of expected dimension for the NDArrays.
```
docstring not being parsed correctly